src/components/elements/upload/Uploader.tsx

```diff
@@ -0,0 +1,393 @@
+            ...existingFiles.filter((f) => f !== file),
+            ...newUploads.map((u) => u.signedBlobId),
+          ]);
+        }
+        if (onUpload) {
+          onUpload(newUploads, newFiles);
+        }
+      } else {
+        if (onChange) {
+          onChange(undefined);
+        }
+        if (onUpload) {
+          onUpload(undefined, undefined);
+        }
+      }
+    },
+    [currentFiles, currentUploads, existingFiles, multiple, onChange, onUpload]
+  );
+
+  const inputRef = useRef<HTMLInputElement | null>(null);
+
+  const showFileList = useMemo(
+    () => multiple && (currentFiles.length > 0 || existingFiles.length > 0),
+    [currentFiles.length, existingFiles.length, multiple]
+  );
+
+  return (
+    <Stack>
+      <Box
+        sx={({ palette, shape }) => ({
+          minHeight: '150px',
+          transition: 'background 300ms',
+          borderRadius: `${shape.borderRadius}px`,
+          border: `1px dashed ${palette.divider}`,
+          // Adjust bottom border if file list is showing (only for multi)
+          ...(showFileList && {
+            borderBottomLeftRadius: 0,
+            borderBottomRightRadius: 0,
+            borderBottom: 0,
+          }),
+          backgroundColor: isDragActive
+            ? alpha(palette.primary.light, 0.12)
+            : palette.background.paper,
+          overflow: 'hidden',
+        })}
+        {...getRootProps()}
+        id={id}
+      >
+        <input
+          {...getInputProps()}
+          ref={inputRef}
+          aria-label={ariaLabel ? ariaLabel : 'Upload file'}
+        />
+        <Stack
+          spacing={1}
+          sx={{
+            justifyContent: 'center',
+            alignItems: 'center',
+            minHeight: '150px',
+            textAlign: 'center',
+            p: 2,
+          }}
+        >
+          {loading && (
+            <Box>
+              <Typography variant='subtitle1' color='inherit' gutterBottom>
+                Uploading
+              </Typography>
+              <LinearProgress variant='indeterminate' />
+            </Box>
+          )}
+          <Box aria-live='polite'>
+            {!loading &&
+              (multiple ||
+                (currentFiles.length === 0 && existingFiles.length === 0)) && (
+                <>
+                  <FilePreviewIcon IconComponent={UploadFileIcon} />
+                  <Typography variant='subtitle1' color='inherit'>
+                    <Link onClick={open} variant='inherit'>
+                      Click to upload
+                    </Link>{' '}
+                    or drag and drop
+                  </Typography>
+                  <Typography variant='body2' color='GrayText' sx={{ mt: 1 }}>
+                    {accept ? getFileTypesFromAccept(accept) : 'Any file type'}{' '}
+                    (max. {getReadableSize(maxSize)})
+                  </Typography>
+                </>
+              )}
+          </Box>
+          <Box aria-live='polite'>
```

Comment:
From here down, there is a lot of logic to reason about for the different scenarios. I would consider refactoring this to reduce the complexity of the render fn component.

I might approach it like this:
```ts
// if it's a single-file upload, render the summary above the error (single thumbnail)
{!multiple && <SingleFileUploadSummary file={currentFiles[0] || existingFiles[0]} onRemove={removeFile} />}}
// upload error
{error}
// if it's a multi-file upload, render the summary below the error (rows)
{multiple && <MultiFileUploadSummary uploadedFiles={[]} stagedFiles={[]} onRemove={removeFile} />}
```
or maybe,

```ts

const fileSummary = (uploadedFiles.length || stagedFiles.length) ?  (<FileUploadSummary uploadedFiles={[]} stagedFiles={[]} onRemove={removeFile} multiple={multiple} />) : null

// if it's a single-file upload, render the summary above the error (single thumbnail)
{!multiple && fileSummary}
// upload error
{error}
// if it's a multi-file upload, render the summary below the error (rows)
{multiple && fileSummary}
```


You can take or leave this feedback. It's very rough. It would help with the complexity of the JSX, but it doesn't help much with reducing the complexity of the component logic overall. There also might be some tradeoffs, like more intermediate components being harder to reason about.

Reply:
Sounds good, I like this idea and took a stab at it here https://github.com/greenriver/hmis-frontend/pull/982/commits/077d08185f8cce3fb996be3864c53742dd99fee8

Reply:
cool, looks good!